### PR TITLE
Feature/complex topic names

### LINKED ISSUE #222: [mqtt] Allow complex topic names in the MQTT Endpoint
Especially for "legacy" applications migrating from an MQTT Broker to Drogue it could be hard to go from a complex topic structure to the "simple" topic structure there currently is.

If we would just allow these topics to still contain their complex structure ("a/b/c/d") but transform them as IS to the subject we could at least make the Produer Side transparent in this regard.

Of course one needs to take special care on the Consumer side as e.g. in case of the MQTT Integration / Consumer users may expect to be able to use patterns on topics (but as far as I have seen you currently can only subscribe on devices and get the full CloudEvents JSON Encoded as payload so it also would not matter in that regard).

### PR DESCRIPTION
**Pull request by JulianFeinauer**
This pull request adds two different MQTT 3.1 topic name parsing strategies.
Currently only the "DrogueV1" is active and there is no way to switch it via configuration just now.

Closes #222.

### CODE REVIEW

**Comment by JulianFeinauer** [comment]
The only thing that is left to be able to finish #222 is to have the possibility to adjust the settings which strategy to use, somewhere. @ctron you have an idea how / where to do that, right?

File: mqtt-endpoint/src/service/session/mod.rs [hunk 1]
```diff
@@ -27,6 +27,50 @@ use std::{
     sync::Arc,
 };
 
+#[derive(Clone)]
+pub enum Dialect {
+    DrogueV1,
+    LongTopicDialect,
```
Comment (ctron) [inline mqtt-endpoint/src/service/session/mod.rs:33]: I would rename this `PlainTopic`. Drop the `Dialect` suffix, as these are all dialects and it just is the plain topic.

File: mqtt-endpoint/src/service/session/mod.rs [hunk 2]
```diff
@@ -108,33 +155,34 @@ where
...
-                        .map_ok(|result| match result.outcome {
-                            GatewayOutcome::Pass { r#as } => Some(r#as),
-                            _ => None,
-                        })
-                })
-                .await
-                .map(|r| (channel.to_string(), r)),
-            _ => Err(PublishError::TopicNameInvalid),
+        match &self.dialect.handle(publish.topic().path()) {
+            Ok(topic) => match topic.device {
```
Comment (ctron) [inline mqtt-endpoint/src/service/session/mod.rs:161]: I think you can merge the two `match` layers into something like:

```rust
match &self.dialect.handle(...) {
  Ok(ParsedTopic{device: Some(device), channel}) => {},
  Ok(ParsedTopic{device: None, channel}) => {}
  Err(_) => {}
}
```

Comment (ctron) [reply mqtt-endpoint/src/service/session/mod.rs:119]: Ok, I just tried that out. It works. But I doesn't make the code more readable. So let's keep it as it is.

**Comment by ctron** [comment]
CI looks good too. @JulianFeinauer let me know if/when you want me to merge this.

**Pull request merged**